packages/endo/src/assemble.js

```diff
@@ -0,0 +1,68 @@
+/* global Compartment */
+
+import { resolve } from "./node-module-specifier.js";
+
+const { entries } = Object;
+
+const defaultCompartment = Compartment;
+
+// q, as in quote, for strings in error messages.
+const q = JSON.stringify;
+
+// Assemble a DAG of compartments as declared in a compartment map starting at
+// the named compartment and building all compartments that it depends upon,
+// recursively threading the modules exported by one compartment into the
+// compartment that imports them.
+// Returns the root of the compartment DAG.
+// Does not load or execute any modules.
+// Uses makeImportHook with the given "root" string of each compartment in the
+// DAG.
+// Passes the given endowments and external modules into the root compartment
+// only.
+export const assemble = ({
+  name,
```

Comment:
This `name` is the same as `packageLocation` in `compartmap.js`, right? And it's not the `label` from that data structure. And I think it's nominally a URL (although `compartments[name].root`, which happens to be a copy of `name`, is probably more deserving of an identifier which has "location" in the name since I think that `.root` will be used to actually find the code).

If the key of `compartments` is not a load-bearing, I mean, location-bearing identifier, then `name` is a great parameter name to use. But maybe there's something we could to in `compartmap.js` to make that clear. Not sure what, maybe:

```js
const translateGraph = (mainPackagePath, graph) => {
  const compartments = {};

  for (const [packageLocation, { label, dependencies }] of entries(graph)) {
  ...
  const name = packageLocation; // not used for location properties, just a distinct name for each compartment
  ...
    compartments[name] = {
      label, // cosmetic, 'packagename@version'
      root: packageLocation, // used to find code for the package
```

Reply:
The for purposes of assembling compartments, the names are just internal references and are only obliged to be internally consistent. In practice, if the compartment map is prepared by importLocation, they are locations; if the compartment map is prepared by writeArchive, they are unique identifiers. Both of these mechanisms are free to change independent of the assembler.

I can probably make that more clear in code as you’re suggesting.